Inhibiting the casting of `x|null` to `?x`  with `null_casts_as_any_type`

morria opened the pull request:
When `Config::get()->null_casts_as_any_type` is enabled, we shouldn't convert parameters with a default value of null to their nullable types.

```php
function f(int $p) {}
function g(int $p = null) {
    f($p);
}
```

In this example, Phan would emit an issue disallowing the casting of `?int` to `int` in the call to `f`, however, if `null_casts_as_any_type`, we'd expect it to work.

TysonAndre commented:
In this example, Phan would emit an issue disallowing the casting of `?int` to `int` in the call to f

Shouldn't change the behavior of comparing `?int` to `int` change, to allow that? (The types that are inferred should remain the same, but the rules for checking if they are compatible should differ.)

- Also, you'd have similar issues for casting `@param ?int $x`

Also, this affects PhanCompatibleSignatureMismatch, but thankfully not RealMismatch

morria commented:
Created issue https://github.com/etsy/phan/issues/699 for tracking this.

morria commented:
@TysonAndre Sounds good. I updated the patch and added a test.

I also modified the testing infrastructure such that a config file can be passed along with the test files to be loaded before running the test and unloaded after.

Let me know if you think this patch is ready to be merged.

morria commented on tests/Phan/MultiFileTest.php:
```diff
@@ -56,6 +56,15 @@ public function getTestFiles() {
+                    MULTI_FILE_DIR . DIRECTORY_SEPARATOR . '699.php',
+                ],
+                MULTI_EXPECTED_DIR . DIRECTORY_SEPARATOR . '699.php' . AbstractPhanFileTest::EXPECTED_SUFFIX,
+                MULTI_FILE_DIR . DIRECTORY_SEPARATOR . '699_config.php',
```
The third element of the array is optional, but if set defines a config file to load when running the test.

TysonAndre commented on tests/Phan/AbstractPhanFileTest.php:
```diff
@@ -147,5 +157,12 @@ public function testFiles($test_file_list, $expected_file_path) {
+
+        // Reset any test-specific config modifiers
+        if ($config_file_path) {
+            foreach ($original_config_value ?? [] as $key => $value) {
```
Phan isn't able to analyze this, could move $original_config_value = `[]` outside of the `if`
tests/Phan/AbstractPhanFileTest.php:163 PhanUndeclaredVariable Variable $original_config_value is undeclared

rlerdorf replied:
We shouldn't emit PhanUndeclaredVariable on the left side of a ??

TysonAndre replied:
It's sort of helpful in many cases in catching typos, and no-ops (Places where the variable was once declared in a method, but no longer is)
But yes, this is a false positive

TysonAndre commented on tests/Phan/AbstractPhanFileTest.php:
```diff
@@ -147,5 +157,12 @@ public function testFiles($test_file_list, $expected_file_path) {
             "Unexpected output in {$test_file_list[0]}");
+
+        // Reset any test-specific config modifiers
+        if ($config_file_path) {
```
This should be done in a `finally` block, or (preferably) tearDown() for this method (Assign to $this->original_config_value)

If a test run with an overridden config fails, then its config would then stay around for other tests, causing confusing test failures with too few issues.

morria commented:
Made a few changes. @TysonAndre Let me know if this is looking good to merge.

TysonAndre commented:
Seems fine
